pkg/datasource/data.go

```diff
@@ -26,15 +26,82 @@ import (
 	"strings"
 	"sync"
 
+	"google.golang.org/protobuf/proto"
+
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
 )
 
-type (
-	data  api.GadgetData
-	field api.Field
-)
+type dataElement api.DataElement
+
+func (d *dataElement) private() {}
+
+func (d *dataElement) payload() [][]byte {
+	return d.Payload
+}
+
+type data struct {
+	*api.GadgetData
+
+	// dataElement is a pointer to the underlying DataElement. It is necessary
+	// to implement the Data interface.
+	*dataElement
+}
+
+func (d *data) SetSeq(seq uint32) {
+	d.Seq = seq
+}
+
+func (d *data) Raw() proto.Message {
+	return d.GadgetData
+}
+
+type dataArray struct {
+	*api.GadgetDataArray
+
+	payloadCount uint32
+}
+
+func (d *dataArray) SetSeq(seq uint32) {
+	d.Seq = seq
+}
+
+func (d *dataArray) Raw() proto.Message {
+	return (*api.GadgetDataArray)(d.GadgetDataArray)
+}
+
+func (d *dataArray) Iterate() *DataIterator {
+	size := len(d.DataArray)
+	elems := make([]Data, 0, size)
+	for _, d := range d.DataArray {
+		elems = append(elems, (*dataElement)(d))
+	}
```

Comment:
I'm not sure about this one - couldn't be just use a reference to `d.DataArray` here? Also, `DataIterator` could be returned as a non-pointer as it's small enough. Maybe it makes sense to do some benchmarks (also comparing it to using `.Count()` and `.GetElement(idx)` on the array directly instead of using the Iterator), but it might also be just premature optimization 😄

Reply:
> I'm not sure about this one - couldn't be just use a reference to `d.DataArray` here? 
>
I don't think so because `DataArrya` is a `[]*api.DataElement`. It means that elements don't implement `Data`. Considering that the iterator will only be used by frontend, we could make this allocation in `SubscribeArray` so that subscribers receive directly a `[]Data`. WDYT?

> Maybe it makes sense to do some benchmarks (also comparing it to using `.Count()` and `.GetElement(idx)` on the array directly instead of using the Iterator), but it might also be just premature optimization 😄
>
IMO, too premature.

Reply:
Let me run some experiments. I think I have an idea.

Reply:
I found a way to avoid creating a new array but work directly with the underlay `DataArray`.